agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/statsbeat/StatsbeatModule.java

```diff
@@ -46,51 +46,90 @@
   private final NetworkStatsbeat networkStatsbeat;
   private final AttachStatsbeat attachStatsbeat;
   private final FeatureStatsbeat featureStatsbeat;
+  private final FeatureStatsbeat instrumentationStatsbeat;
 
   private final AtomicBoolean started = new AtomicBoolean();
+  private final AtomicBoolean disabledAll = new AtomicBoolean();
+
+  private volatile TelemetryClient telemetryClient;
 
   private StatsbeatModule() {
     customDimensions = new CustomDimensions();
     networkStatsbeat = new NetworkStatsbeat(customDimensions);
     attachStatsbeat = new AttachStatsbeat(customDimensions);
-    featureStatsbeat = new FeatureStatsbeat(customDimensions);
+    featureStatsbeat = new FeatureStatsbeat(customDimensions, FeatureType.Feature);
+    instrumentationStatsbeat = new FeatureStatsbeat(customDimensions, FeatureType.Instrumentation);
   }
 
   public void start(TelemetryClient telemetryClient, Configuration config) {
     if (started.getAndSet(true)) {
       throw new IllegalStateException("initialize already called");
     }
 
-    long intervalSeconds = config.internal.statsbeat.intervalSeconds;
-    long featureIntervalSeconds = config.internal.statsbeat.featureIntervalSeconds;
-
-    scheduledExecutor.scheduleWithFixedDelay(
-        new StatsbeatSender(networkStatsbeat, telemetryClient),
-        intervalSeconds,
-        intervalSeconds,
-        TimeUnit.SECONDS);
-    scheduledExecutor.scheduleWithFixedDelay(
-        new StatsbeatSender(attachStatsbeat, telemetryClient),
-        intervalSeconds,
-        intervalSeconds,
-        TimeUnit.SECONDS);
-    scheduledExecutor.scheduleWithFixedDelay(
-        new StatsbeatSender(featureStatsbeat, telemetryClient),
-        featureIntervalSeconds,
-        featureIntervalSeconds,
-        TimeUnit.SECONDS);
-
-    ResourceProvider rp = customDimensions.getResourceProvider();
-    // only turn on AzureMetadataService when the resource provider is VM or UNKNOWN.
-    // FIXME (heya) Need to figure out why AzureMetadataService is not reachable from a function app
-    // and it's not necessary to make this call.
-    if (rp == ResourceProvider.RP_VM || rp == ResourceProvider.UNKNOWN) {
-      // will only reach here the first time, after instance has been instantiated
-      new AzureMetadataService(attachStatsbeat, customDimensions)
-          .scheduleWithFixedDelay(intervalSeconds);
+    if (config.internal.statsbeat.disabledAll) {
+      // disabledAll is an internal emergency kill-switch to turn off Statsbeat completely when
+      // something goes wrong.
+      // this happens rarely.
+      disabledAll.set(true);
+      return;
     }
 
-    featureStatsbeat.trackConfigurationOptions(config);
+    if (!config.internal.statsbeat.disabled) {
+      if (this.telemetryClient == null) {
+        this.telemetryClient = telemetryClient;
```

Comment:
it looks like `telemetryClient` is not initialized when non-essential statsbeat are disabled?